derive/src/lib.rs

```diff
@@ -27,48 +28,167 @@ pub fn contract(input: proc_macro::TokenStream) -> TokenStream {
 /// Contract procedural macro arguments.
 struct ContractArgs {
     artifact_path: LitStr,
-    builder: Builder,
+    parameters: Vec<Parameter>,
 }
 
-// TODO(nlordell): Due to limitation with the proc-macro Span API, we can't
-//   currently get a path the the file where we were called from; therefore,
-//   the path will always be rooted on the cargo manifest directory.
-//   Eventually we can use the `Span::source_file` API to have a better
-//   experience.
+impl ContractArgs {
+    fn into_builder(self) -> Builder {
+        let mut builder = Builder::new(&self.artifact_path.value());
+        for parameter in self.parameters.into_iter() {
+            builder = match parameter {
+                Parameter::Crate(name) => builder.with_runtime_crate_name(name),
+                Parameter::Deployments(deployments) => {
+                    deployments.into_iter().fold(builder, |builder, d| {
+                        builder.add_deployment(d.network_id, d.address)
```

Comment:
That's a good point. I will look into propagating compiler warnings so they see this. The idea here would be if you do this:
```
ethcontract::contract!(
    "path/to/artifact",
    deployments {
        1 => "0x...",
        2 => "0x...",
        1 => "0x...", // should be a compiler warning on this line
    }
);
```

Reply:
Currently, Rust doesn't support warnings from proc_macros in stable (there is support on nightly, but I don't think it will be stabilized anytime soon). This means its not possible to warn when a manually specified contract address collides with a JSON artifact one.

I think that for now, overwriting the artifact's network configuration is the correct behaviour given the limitations. The idea of manually specified deployments is to be able to:
- provide deterministic addresses for testing scenarios with development nodes (like what we want for `dex-services` with Ganache tests)
- provide a way to override addresses from a truffle artifact (for example if we are generating bindings to a contract that has its artifact published on NPM but have our own deployment to the contract that we want to use instead).

The explicitness of the `deployments { ... }` parameter in the macro invocation makes be not feel too bad about it not being an error.

I did add a hard compile error in the case where a deployment gets specified more than once in the proc macro invocation:
```
error: duplicate network ID in `ethcontract::contract!` macro invocation
 --> examples/deployments.rs:9:9
  |
9 |         5777 => "0x0123456789012345678901234567890123456789",
  |         ^^^^
```

@fleupold Let me know if you are OK with this.

Reply:
One idea I just had was to slightly modify the syntax to be like this:
```
5777! => "0x..." // override network ID 5777
5777 => "0x..." // error if already exists in contract.
```

If you think that is a good alternative, I'll open a new issue and do it in a separate PR.

Reply:
I think that’s a very neat idea. Wondering if it should only be an error if the address in the contract is different?

Reply:
Created #130